### ISSUE TITLE
at-xy

### ISSUE THREAD

**Issue by lonetech** [jkotlinski/durexforth#438]
I didn't see where in the library [at-xy](https://forth-standard.org/standard/facility/AT-XY) was, so I cobbled one together using [kernal plot](https://www.c64-wiki.com/wiki/PLOT_(KERNAL)):

    : at-xy xr c! yr c! 0 sr c! $fff0 sys ;

Do you think it might be worth including somewhere? 

It doesn't preserve interrupt disable, which might be something to improve. In assembly it would use CLC.

The routine itself would be grouped with page and emit, so I guess the norm here is assembly routines in io.asm. 
I did the exercise to convert to `code`:

    code at-xy ( x y -- )
      clc,
      w stx,
      \ w 1+ sty,
      lsb 1+ ldy,x
      lsb lda,x  \ had no ldx,x
      tax,
      $fff0 jsr,
      w ldx,
      \ w 1+ ldy,
      inx,
      inx,
    ;code

That also worked. It's rather cumbersome code, though, mostly dealing with ABI conversion. Still `size` reports the code version as smaller than the sys version. 

Evidently, I am a total newbie at 6502 assembly.

**Comment by jkotlinski** [comment]
Nice! Maybe it would fit in compat.fs as an optional include?

**Comment by lonetech** [comment]
It does fit the compat.fs pattern as a word with standard functionality that isn't currently in use. I'd be quite happy with that.

**Comment by Whammo** [comment]
Perhaps it's corollary would also be appropriate? XY-AT ? XY@ ? PLOT ?

**Comment by Whammo** [comment]
```
: setcur ( x y -- )  \ hidden in v
xr ! yr ! $e50c sys ;

vs.

code setcur ( x y -- )
   lsb ldy,x
       inx,
   lsb lda,x
       inx,
     w stx,
       tax,
 $e50c jsr,
     w ldx,
;code


```

**Comment by lonetech** [comment]
That's jumping inside the PLOT routine, to the set part. The $fff0 vector points to $e50a, which presumably does a branch if the carry flag is set (bcs) before falling through to $e50c. This sort of shortcut might make it incompatible with e.g. Commodore 128 80-column mode.

**Comment by Whammo** [comment]
Fun, fun, fun!

```forth
: setcur ( x y -- )  \ hidden in v
xr ! yr ! $e50c sys ;

code at-xy ( x y -- )
   lsb lda,x
       inx,
   lsb ldy,x
       inx,
     w stx,
       tax,
 $e50c jsr,
     w ldx,
;code

size setcur \ 20
size at-xy  \ 15
```

**Comment by jkotlinski** [comment]
A pull request with this addition would be welcome. I suppose it should be added to documentation and changelog, too.

**Comment by Whammo** [comment]
This doesn't set the screen pointers properly,

```: goto-xy ( x y – ) 214 c! 211 c! ;```

**Comment by ZbigniewB** [comment]
Alas, you're right; X works OK, but Y (214) seems to make sense just as read-only. Probably during my quick-check I used the value that moved cursor into the current line and on first glance it looked OK.

**Comment by Whammo** [comment]
It's great fun to jump into the Kernal.
   
```: goto-xy ( x y – ) 214 c! 211 c! $e56c sys ; ```

**Comment by ZbigniewB** [comment]
Indeed. I have to admit my second mistake: since Durex Forth strives to be ANS-standard compliant, that word has to be named AT-XY, unfortunately, not GOTO-XY, But my new proposal for the name of „a word, that returns character code at X, Y” is XY@ ( x y -- char ) (on PC it would return character code and attribute byte).

**Comment by ZbigniewB** [comment]
Its code will be rather simple:
XY@ ( x y -- charcode ) 40 * + 53272 C@ 16 / 1024 * + C@ ;

**Comment by Whammo** [comment]
It is fun to not presume from where screen memory may be obtained.  : )

**Comment by farvardin** [comment]
I agree it should be included in at least the compat.fs file. Both solutions works great, so either

`: at-xy xr c! yr c! 0 sr c! $fff0 sys ;
`
or
`
: at-xy ( x y – ) 214 c! 211 c! $e56c sys ;`

with for example this code:

```
\ https://logiker.com/Vintage-Computing-Christmas-Challenge-2023
\ https://ftp.scene.org/pub/parties/2023/vccc23/results.txt
\ code by Kempston

: at-xy xr c! yr c! 0 sr c! $fff0 sys ;
: at-xy ( x y – ) 214 c! 211 c! $e56c sys ;

: diamond
0
page
begin
dup dup 19 / swap
19 mod over over 18
swap - at-xy 42 emit
at-xy 42 emit 6 +
dup 361 >
until
cr
;

diamond
```

<img width="602" height="632" alt="Image" src="https://github.com/user-attachments/assets/dd4d9da2-0847-4e76-ab0d-72498ba655eb" />
